applications/ShapeOptimizationApplication/custom_utilities/optimization_utilities.h

```diff
@@ -152,6 +152,105 @@ class KRATOS_API(SHAPE_OPTIMIZATION_APPLICATION) OptimizationUtilities
         LinearSolver<DenseSpace, DenseSpace>& rSolver,
         Vector& rProjectedSearchDirection,
         Vector& rRestoration);
+
+     // ==============================================================================
+    // For running relaxed gradient projection
+    // ==============================================================================
+/**
+     * Assemble a list of Numbers into a Vector, independent of the model part
+     */
+    static void AssembleBufferVector( Vector& rVector,
+        const std::vector<double>& rVariables)
+    {
+    	size_t VectorSize = rVariables.size();
+        if (rVector.size() != VectorSize){
+            rVector.resize(VectorSize);
+        }
+
+
+        for (size_t i=0;i < VectorSize;i++)
+        {
+            rVector[i] = rVariables[i];
+        }
+    }
+    /**
+     * Assemble a list of Numbers into a diagonal Matrix, independent of the model part
+     */
+    static void AssembleBufferMatrix( Matrix& rMatrix,
+        const std::vector<double>& rVariables)
+    {
+    	size_t VectorSize = rVariables.size();
+        rMatrix = ZeroMatrix(VectorSize, VectorSize);
+        for (size_t i=0; i < VectorSize; i++)
+        {
+            rMatrix(i,i) = rVariables[i];
+        }
+    }
+    /**
+     * Assemble a list of Vectors into a  Matrix
+     */
+    static void AssembleVectorstoMatrix(ModelPart& rModelPart,
+        Matrix& rMatrix,
+        const std::vector<Vector*>& rVariables
+    )
+    {
+        if ((rMatrix.size1() != rModelPart.NumberOfNodes()*3 || rMatrix.size2() !=  rVariables.size())){
+            rMatrix.resize(rModelPart.NumberOfNodes()*3, rVariables.size());
+        }
+
+
+    	int j=0;
+    	for (Vector* p_variable_j : rVariables)
+    	{
+        	const Vector& r_variable_j = *p_variable_j;
+
+
+        	for (size_t i = 0; i < rModelPart.NumberOfNodes()*3; i++)
+        	{
+        		rMatrix(i, j) = r_variable_j[i];
+        	}
+   		++j;
+ 	}
+
+    }
```

Comment:
this also can be parallelized for OMP

Reply:
removed and the existing assembleMatrix is used.